lib/maths/unittest/CTreeShapFeatureImportanceTest.cc

```diff
@@ -0,0 +1,189 @@
+
+        treeFeatureImportance =
+            std::make_unique<maths::CTreeShapFeatureImportance, std::initializer_list<TTree>>(
+                {tree1, tree2});
+        CStubMakeDataFrameCategoryEncoder stubParameters{1, *frame, 0};
+        encoder = std::make_unique<maths::CDataFrameCategoryEncoder>(&stubParameters);
+    }
+
+    ~SFixtureMultipleTrees() {}
+
+    TDataFrameUPtr frame;
+    std::size_t numberFeatures{2};
+    std::size_t numberRows{10};
+    TTreeShapFeatureImportanceUPtr treeFeatureImportance;
+    TEncoderUPtr encoder;
+};
+
+BOOST_FIXTURE_TEST_CASE(testSingleTreeSamplesPerNode, SFixtureSingleTree) {
+
+    auto samplesPerNode = maths::CTreeShapFeatureImportance::samplesPerNode(
+        treeFeatureImportance->trees()[0], *frame, *encoder, 1);
+    TDoubleVec expectedSamplesPerNode{4, 2, 2, 1, 1, 1, 1};
+    BOOST_TEST_REQUIRE(samplesPerNode == expectedSamplesPerNode);
+}
+
+BOOST_FIXTURE_TEST_CASE(testSingleTreeExpectedNodeValues, SFixtureSingleTree) {
+
+    TDoubleVec samplesPerNode{4, 2, 2, 1, 1, 1, 1};
+    std::size_t depth = maths::CTreeShapFeatureImportance::updateNodeValues(
+        treeFeatureImportance->trees()[0], 0, samplesPerNode, 0);
+    BOOST_TEST_REQUIRE(depth == 2);
+    TDoubleVec expectedValues{10.5, 5.5, 15.5, 3.0, 8.0, 13.0, 18.0};
+    auto& tree{treeFeatureImportance->trees()[0]};
+    for (int i = 0; i < tree.size(); ++i) {
+        BOOST_TEST_REQUIRE(tree[i].value() == expectedValues[i]);
+    }
+}
+
+BOOST_FIXTURE_TEST_CASE(testSingleTreeShapNotNormalized, SFixtureSingleTree) {
+    std::size_t offset{frame->numberColumns()};
+    treeFeatureImportance->shap(*frame, *encoder, frame->numberColumns(), offset);
+    TDoubleVecVec expectedPhi{{-5., -2.5}, {-5., 2.5}, {5., -2.5}, {5., 2.5}};
+    frame->readRows(1, [&](TRowItr beginRows, TRowItr endRows) {
+        for (auto row = beginRows; row != endRows; ++row) {
+            for (std::size_t col = 0; col < 2; ++col) {
+                BOOST_TEST_REQUIRE((*row)[offset + col] == expectedPhi[row->index()][col]);
+            }
+        }
+    });
+}
+
+BOOST_FIXTURE_TEST_CASE(testMultipleTreesShapNotNormalized, SFixtureMultipleTrees) {
+    TDoubleVecVec expectedPhi{
+        {-0.82660001, -0.06222489}, {-0.82660001, -0.06222489},
+        {-0.82660001, -0.06222489}, {-0.58498581, -0.06222489},
+        {-0.58498581, -0.06222489}, {0.03993395, -0.06222489},
+        {0.90245943, -0.2177871},   {1.0269092, 0.0725957},
+        {1.0269092, 0.0725957},     {1.0269092, 0.0725957}};
+    std::size_t offset{frame->numberColumns()};
+    treeFeatureImportance->shap(*frame, *encoder, frame->numberColumns(), offset);
+    frame->readRows(1, [&](TRowItr beginRows, TRowItr endRows) {
+        for (auto row = beginRows; row != endRows; ++row) {
+            for (std::size_t col = 0; col < 2; ++col) {
+                BOOST_TEST_REQUIRE((*row)[offset + col] == expectedPhi[row->index()][col]);
+            }
+        }
+    });
+}
```

Comment:
I think a nice test would be to test invariant "sum of shap values" = "prediction - mean" for some randomised input.

Reply:
Done.